File: tests/bench_marks/cpp/07_fibonacci_memoized.cpp
// ベンチマーク7: フィボナッチ数列（動的計画法版）
// 45番目のフィボナッチ数を効率的に計算

#include <chrono>
#include <iostream>

using namespace std;
using namespace std::chrono;

int main() {
    int n = 45;  // int範囲内に収まる値

    auto start = high_resolution_clock::now();

    // 動的計画法でフィボナッチを計算
    if (n <= 0) {
        cout << "Fibonacci(" << n << ") = 0" << endl;
        return 0;
    }

    if (n == 1) {
        cout << "Fibonacci(" << n << ") = 1" << endl;
        return 0;
    }

    int prev2 = 0;
    int prev1 = 1;
    int result = 0;

    for (int i = 2; i <= n; i++) {
        result = prev1 + prev2;
        prev2 = prev1;
        prev1 = result;
    }

    auto end = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(end - start);

    cout << "Fibonacci(" << n << ") with DP = " << result << endl;
    cout << "Time: " << duration.count() / 1000000.0 << " seconds" << endl;

    return 0;
}

File: tests/bench_marks/rust/07_fibonacci_memoized.rs
// ベンチマーク7: フィボナッチ数列（動的計画法版）
// 45番目のフィボナッチ数を効率的に計算

use std::time::Instant;

fn main() {
    let n = 45;  // int範囲内に収まる値

    let start = Instant::now();

    // 動的計画法でフィボナッチを計算
    if n <= 0 {
        println!("Fibonacci({}) = 0", n);
        return;
    }

    if n == 1 {
        println!("Fibonacci({}) = 1", n);
        return;
    }

    let mut prev2 = 0;
    let mut prev1 = 1;
    let mut result = 0;

    for _ in 2..=n {
        result = prev1 + prev2;
        prev2 = prev1;
        prev1 = result;
    }

    let duration = start.elapsed();

    println!("Fibonacci({}) with DP = {}", n, result);
    println!("Time: {:.6} seconds", duration.as_secs_f64());
}